#3 documentation front end (#80)
openmsupply/application-manager-web-app: documentation/internal/Development-Workflow.md

_This should help Front-end developers understand the flow of how React components and GraphQL calls interact and the current project structure._

# Project Structure

We have a few rules to maintain our structure comprehensive. This rules may change with time, and we should keep this file up-to-date with the current ones. Check the rules in each section below.
Inside the `/src` folder all files used for the Front-end are distribuited in the following structure:

- `components`
- `containers`
- `contexts`

> - `elememtPlugins`

Reviewer: ```suggestion
- `elementPlugins`
```

- `utils`
  - `generated`
  - `graphql`
  - `helpers`
  - `hooks`

### components

`Presentation components` or functional UI React components.

> **Rule 1**: Not logic or should be done at the component level, it simply renders what is received in the `props`.

Reviewer: ```suggestion
**Rule 1**: No logic should be done at the component level, it simply renders what is received in the `props`.
```

> **Rule 1**: Not logic or should be done at the component level, it simply renders what is received in the `props`.

Author: Yes, I think we need this rule to help us differentiate what is a `container` and a  `component`. Some logic can still be done in the component, specially if that's done while developing something, but we should keep all calls to hooks and management of cache in the controllers.

**Rule 1**: Avoid using context for anything that can be fetched from the database (using Apollo client) or set in the local state (`useState`) and passed to other components as `props`.

- Apollo client CACHE is doing most part of the local state management, other local states can be shared between components by `props`. But a few times we need to be passing information (that aren't in the database) from one component to another not directly connect.
- We are studying the idea of using Apollo cache directives (@client) for replacing with the existing contexts states. Or simple local states.
- **Important**: For a component be able to access something from a context area it will need to be (or some parent component needs to be) wrapped by the context provider! Check out how it's done in the `SiteLayout.tsx` file.

### elementsPlugins

aka `Question plugins` - This is used for all related code for the Question plugins.

- Each folder inside will defined each Question-type for the App. Inside each question-type folder there are 2 files:
  - `ApplicationView` => Define how this question is rendered during an application in progress/edited.

>   - `TemplateView` => Defines how this question is rendered during the configuration of the template of one application.

Author: Yes. Indeed. I'll add the Summary View also in there. Thanks

### utils

To keep everything else :P

#### generated

Folder with a single file `graphql.tsx`

- The file gets re-generated (by manually running the script `yarn generate`) with the complete list of all the types related to the tables, relations and fields in the GraphQL server. And adds types to manually.
- The file also gets generated types for the `queries`, `mutation` and `fragments` defined in the next folder...

#### graphql

Folder storing all GraphQL queries, mutation and fragments (so everthing is in one place).

**Important**: Always check that the type for the files included in this folder are using `.ts` - or generation will fail!

**Rule 1**: All GraphQl calls are grouped in here to keep all-in-one-place and to be used on the the type generation.

> **Rule 2**: After adding, creating run the `yarn generate` script.

Reviewer: ```suggestion
**Rule 2**: After adding or editing queries, run the `yarn generate` script.
```

After the generation runs a new **custom hook** with all required types is defined in the `generated/graphql.ts` file named as the pattern.

Example: a new query called `getApplications.query.ts` will get a new hook (for querying) as `useGetApplicationsQuery`.

**Rule 3**: Use the **Pattern**:

- Query: `get` `Entity` `.` `query` `.ts`
- Mutation: (`create` | `update`) `Entity` `.` `mutation` `.ts`
- Fragment: (`add` | `remove`) `Entity` `.` `fragment` `.ts`

##### fragments

> _What is a fragment?_ A partial query or mutation that can be used as part of other queries or (Mainly) for getting or setting some data to the cache. We do have to update the cache after a mutation to keep it up-to-date.

Reviewer: ```suggestion
_What is a fragment?_ A partial query or mutation that can be used as part of other queries or (mainly) for getting or setting some data to the cache. We do have to update the cache after a mutation to keep it up-to-date.
```

##### queries

_What is a query?_ Something to be fetched from the GraphQL server. Can have some input and defines what is returned.

##### mutation

> _What is a mutation?_ Something to be posted to the GraphQL server. Habe some defined input and also something that is returned.

Reviewer: ```suggestion
_What is a mutation?_ Something to be posted to the GraphQL server. Has some defined input and also something that is returned.
```

#### helpers

Add here any methods that can be used in other places - that aren't hooks or compnents itselves.

#### hooks

Our **custom hooks** - exaplained in more detail in the Development sections below.

### Files on the root src folder

- `cache.tsx` - Start the Apollo Cache.

>   Also define some local state to be used within the cache on the root level or as part of any table (@client). Read more [here](https://www.apollographql.com/docs/react/local-state/managing-state-with-field-policies/#storing-local-state-in-the-cache).

Reviewer: ```suggestion
  Also can define some local state to be used within the cache on the root level or as part of any table (@client). Read more [here](https://www.apollographql.com/docs/react/local-state/managing-state-with-field-policies/#storing-local-state-in-the-cache).
```

- `config.json` - Some global varibles to be used during the configuration of the App. i.e: GraphQL server host.

> - `index.tsx` - App start - Just connects the Apollo Cache (from `cache.tsx`), add semantic-ui for UI and call the main component: `App.tsx`.

Reviewer: ```suggestion
`index.tsx` - App start - Just connects the Apollo Cache (from `cache.tsx`), adds semantic-ui for UI and calls the main component: `App.tsx`.
```

---

# Local vs. Remote states

### Remote states

> The local data kept in the app by Apollo client CACHE is considered remote state since is a small mirror of the GraphQL server. The cached data is reused for up-coming queries (if we don't enforce a `refetch` from the server), just returning the local **remote state** for the queried data - when it is available. Or get from the server if not available in cache.

Reviewer: ```suggestion
The local data kept in the app by Apollo client CACHE is considered remote state since it is a small mirror of the GraphQL server. The cached data is reused for up-coming queries (if we don't enforce a `refetch` from the server), returning the local **remote state** for the queried data - when it is available. Or it fetches from the server if not available in cache.
```

**Important**: No need to do anything different to query from the cache instead of the server. Just a simple query will do it. Mutations on the other hand require some addition - after sending data to the server we have to update the cache.

_How Apollo keeps the cache synced with the server?_

> Apollo client provides this cache management module, which will store all the data query or mutation results after it is fetched from the server. It requires you add the `id` of each entity - and the related entities included in the fetched data - or another identifier that you have specified before to the cache management instance if not using the `id`. The caches stores one instance of each entity queried (refreshing would clear the cache). And for each relationship it keeps a reference to its entity. In the end the remote states has one instance of each entity already fetched from the server.

Reviewer: So it doesn't need unique global id? Table and id is sufficient for it to be identified?

Author: Yes. The id should be unique between entities in the same table, that's why by default it uses the table `id` field. :)

_Why do we update the cache after mutations?_

> When we create a new entity that has a relationshio with some existing in cache we need to also update the cache with the added entity! This is to avoid having to create a `refetch` to get this back from the server and keep the local cache up-to-date. The same way, after editing an entity which is already in cache we need to remember to update this entity in the cache.

Reviewer: ```suggestion
When we create a new entity that has a relationship with some existing data in cache we need to also update the cache with the added entity! This is to avoid having to create a `refetch` to get this back from the server and keep the local cache up-to-date. In the same way, after editing an entity which is already in cache we need to remember to update this entity in the cache.
```

##### Example of remote data

- Current Application

> - Application Sections (related enetities - Application)
>   - Template Sections (related enetities - Application Section)

Reviewer: ```suggestion
- Application Sections (related entities - Application)
  - Template Sections (related entities - Application Section)
```

  - Template element (related entities - Application Response)

### Local states

We are using 2 options for local states:

- local variables (with `useState`) passed as `props` to other components
- context areas (with `useContext`) for sharing between components

##### Example of local state

**TODO**: Rename ApplicationPage => ApplicationPageWrapper (if we decide that we want to keep the wrapper thingy).

**To be created** => In ApplicationPageWrapper:

- List of all questions and responses

How to use:

1. `ApplicationPage` calls initial query to load the whole application, section and responses.
2. `ApplicationPage` calls a custom hook to create the list of questions and responses based on query.
3. Custom hook returns the result object Q&R with all the question elements with reduced fields (after running the `evaluateExpression` on each question) and the responses by the user.

The object returned has for each Q&R (Question and Response):

- Question code (key)
- Type: Information | Question - if information none of the next fields are filled!
- isRequired: boolean
- isVisible: boolean
- isValid: boolean
- ResponseId: number
- ResponseJSON: JSON

> - ResponseValue: string

Reviewer: After doing [this](https://github.com/openmsupply/application-manager-server/pull/95) in response to [this](https://github.com/openmsupply/application-manager-server/issues/61), and working on this local state Q & R object, I think the full list of fields will be:

- Question code (key)
- Category: Information | Question
- isRequired: boolean
- isVisible: boolean
- isValid: boolean
- isEditable: boolean
- validationMessage: string
- ResponseId: number
- ResponseJSON: JSON
- ResponseValue: string

Also, regarding your comment "If information none of the next fields are filled" -- `isVisible` can still be relevant!

Author: I have updated this, also talking about 2 objects that should be returned by this custom hook.

4. The Q&R object is passed to `ElementsBox` and each related question and response to the `ApplicationViewWrapper`.
5. The `ApplicationViewWrapper` receives the mutation (from `ElementsBox`) to update the response `onFieldChage`. It should only update to the server and CACHE when the response is valid.

> 6. The `ApplicationPage` will update the Q&R object when the query for elements and responses gets triggerdby the new mutation of one of the elements response on the current application.

Reviewer: ```suggestion
6. The `ApplicationPage` will update the Q&R object when the query for elements and responses gets triggerd by the new mutation of one of the elements' response on the current application.
```

##### Example of context state

`src/contexts/ApplicationState.tsx`

This context area is used between components to share a few states:

- Application serial number (currently created on Front-end)
- Application id
- Application's sections total of pages - which gets deduced by the elements in each section.

---

# Development workflow

## Example

Process of creating one Application, going through each page until reach the summary page where the user can submit.

> ![development_workflow_applications](images/Development_workflow_applications.png)

Reviewer: Image is great. Just one thing: I spent ages wondering what on earth "bt" was. Then I realise you meant "button". If there's no room for the full word "button", maybe just lose the "bt". It's not an abbreviation I've seen before and the diagram box looks enough like an actual button to get the idea.

Author: Hahhaa Sorry! I'll just drop it then : D

  - These components should received (as props or importing one context area) everything to be presented.
  - Any action triggered by button should also be passed to the components.
- Add the new page to the `SiteLayout.tsx` in `src/containers/Main` folder.

  - If you need to have access to a context area, remember to wrap the Router of your container with it.
    Example:

  ```
  <Route exact path="/example">
    <UserProvider>
      <UserRegister />
    </UserProvider>
  </Route>
  ```

- Mainly we are using the Apollo cache for **state managment** (the alternative is to use context areas - the aim is to only use Apollo cache at some point). To gather what needs to be displayed you either will get it from the cached data (of previous calls to GraphQL) or make a new call to the GraphQL server.
  - Apollo client will handle this for you by simply making a call with your desired `query`. [Read more here](https://www.apollographql.com/docs/react/data/queries/#caching-query-results)
  - In some areas we would like to JUST store the changes in cache, so it's required to use `writeQuery` instead.

### Create custom hooks

When starting to create a query first check:

> - If there is already one similar in `utils/graphql/queries` - you can addapt this one - considering other callers.

Reviewer: ```suggestion
- If there is already one similar in `utils/graphql/queries`, you can adapt this one - considering other callers.
```

> - Very similarly check `utils/graphql/mutations` if you need to send data (mutation) to Graphql instead.

Reviewer: ```suggestion
- Similarly, check `utils/graphql/mutations` if you need to send data (mutation) to GraphQL.
```

> - If there isn't one already - Create a new file that just return the query (using the Graphql query language).

Reviewer: ```suggestion
- If there isn't one already, create a new file that just returns the query (using the GraphQL query language).
```

  - The query will also defined what is returned!
  - Use a GUI tool to see the results for your query/mutation to the GraphQL server: `http://localhost:5000/graphiql` (if the server is running on localhost:5000).
  - Another helper is using the extension described in [here](setup/Apollo-clinet.md) on sections "VS code extension for typing GraphQL" to have autocomplete on the query you are writing directly on your file.

> After you have finished to add the query/mutation:

Reviewer: ```suggestion
After you have finished adding the query/mutation:
```

- Run the script `yarn generate`.

  - If all goes well, in the `utils/generated/graphql.ts` file you can now view a **generated hook** for your new query/mutation. It links the Apollo hook `useQuery` or `useMutation` and adds all predefined types from Graphql.

    Example: If your query was called `getUsers.query.ts` the **generated hook** will be called `useGetUsersQuery`.

You just need another **custom hook** created now.

> - The reason to create - yet another - hook is that simply calling the new **generated hook** it's still needed to make some transformation with the data, sometimes even run sequential queries/mutations. The way hooks work, it's better to have another hook calling the generated one so the `container` call just you **custom hook** and get's the final result.

Reviewer: ```suggestion
- The reason to create (yet another) hook is that simply calling the new **generated hook** you still need to make additional transformations of the data, sometimes even run sequential queries/mutations. The way hooks work, it's better to have another hook calling the generated one so the `container` calls just your **custom hook** and gets the final result.
```

##### queries

_What is a query?_ Something to be fetched from the GraphQL server. Can have some input and defines what is returned.

##### mutation

_What is a mutation?_ Something to be posted to the GraphQL server. Has some defined input and also something that is returned.

#### helpers

Add here any methods that can be used in other places - that aren't hooks or compnents itselves.

#### hooks

Our **custom hooks** - exaplained in more detail in the Development sections below.

### Files on the root src folder

- `cache.tsx` - Start the Apollo Cache.
  Also can define some local state to be used within the cache on the root level or as part of any table (@client). Read more [here](https://www.apollographql.com/docs/react/local-state/managing-state-with-field-policies/#storing-local-state-in-the-cache).

- `config.json` - Some global varibles to be used during the configuration of the App. i.e: GraphQL server host.

`index.tsx` - App start - Just connects the Apollo Cache (from `cache.tsx`), adds semantic-ui for UI and calls the main component: `App.tsx`.

---

# Local vs. Remote states

> ### Remote states

Reviewer 2: Would be good to include this diagram.

https://app.diagrams.net/#G17Lj82915_i6Ou6INh92ye-PvYsCjOwRn

![Application Flow](https://user-images.githubusercontent.com/26194949/98053704-10c5da00-1e9e-11eb-9045-8ff7caf22124.png)

The local data kept in the app by Apollo client CACHE is considered remote state since it is a small mirror of the GraphQL server. The cached data is reused for up-coming queries (if we don't enforce a `refetch` from the server), returning the local **remote state** for the queried data - when it is available. Or it fetches from the server if not available in cache.

> **Important**: No need to do anything different to query from the cache instead of the server. Just a simple query will do it. Mutations on the other hand require some addition - after sending data to the server we have to update the cache.

Author: Great I'll add this in.
